### DISCUSSION TITLE
Lora sx1278 spi support for arduino pro mini or arduino nano

### DISCUSSION THREAD

**Discussion by shooka90** [jgromes/RadioLib#246] [Q&A] [answered]
Hi, i have a question:

I have been trying to receive data from another lora sx1278 module, using my arduino nano or arduino pro mini 5v,16mhz (i have both), but i can't receive any data. I am using the example code (SX1278-->Receive with interrupt), but no luck with my arduino pro mini or arduino nano. I have tried the code with an arduino uno, and everything works perfectly. But it seems the code won't work for arduino nano or pro mini. Every time I try it with pro mini or nano i get the error code -16 or -2. I have checked the wiring and everything is correct. I have tried different nanos and pro minis and the problem still exist. 
There is a solution for arduino nano which i have tried and worked. The bootloader on my nano was the old bootloader, I have burned the arduino uno bootloader on the nano, and all of my problems have vanished. The code works and I could receive my data using my arduino nano.
But still no luck with pro mini. I can't burn the uno bootloader on a pro mini.

I cant figure it out for the pro mini. Do you have any suggestions? Am i missing something?

[mylorareceiver.txt](https://github.com/jgromes/RadioLib/files/5946186/mylorareceiver.txt)

**Comment by jgromes** [comment]
From the point of view of the library, there should be no difference between Arduino Uno, Pro Mini and Nano - they're all using ATmega328P MCU. The bootloader should also make no difference at this level, since that's just a piece of code that's used to upload the program via Serial connection - but if your Nano/Uno/Pro Mini is a Chinese clone off Ebay, then I suppose there's really no way to tell what exactly is in there.

Digging through Arduino AVR boards.txt, the only two differences I can see between Uno and Nano/Pro Mini are a slight change in one fuse setting (which sets bootloader size) and the bootloader file that's used.

Can you post more information about this? Debug mode dumps from Nano and Pro Mini, full build output (after enabling verbose output in Arduino IDE during compilation), versions of everything used (the library, Arduino IDE version, Arduino AVR boards version),... - basically anything that could help to narrow this down.

Also, assuming your LoRa module has 3.3V logic (most do), are you using a voltage level shifter where appropriate?

**Comment by shooka90** [comment, marked as answer]
I think I am using clone pro mini. I have powered on my arduino pro mini with 3.3v and upload my code in it (so technically I am just overclocking it). So, now I dont need a level shifter (i guess). I have compiled and upload the code and still no luck. I have provided the verbose and attached the it to this post.
Again, I have tried my code again with arduino nano (and no level shifter), and everythings is fine. I have provided the verbos of nano in the attachment. 

I have also uploaded my latest code.
[thecode.txt](https://github.com/jgromes/RadioLib/files/5950189/thecode.txt)


Is there anything else you need to narrow it down?

[verbos_nano.txt](https://github.com/jgromes/RadioLib/files/5950102/verbos_nano.txt)
[verbos_promini5v16mhz.txt](https://github.com/jgromes/RadioLib/files/5950103/verbos_promini5v16mhz.txt)

**Reply by jgromes** [reply]
>  have tried my code again with arduino nano (and no level shifter), and everythings is fine

Nano is a 5 V device, so in the long run, things most likely won't be fine for the LoRa module.

The only thing that stands out in the build logs is that you're building for Arduino Pro Mini @ 8 MHz, not 16, that's probably not relevant.

I can see you're using Arduino IDE 1.8.13, unfortunately the version of Arduino AVR core isn't included in the build log - could you check it in the Boards Manager? It should be the newest one (1.8.3). I'll se if I can replicate this on my end, though I don't recall having any issues with neither Nano or Pro Mini in the past.

Also, could you post [debug mode](https://github.com/jgromes/RadioLib/wiki/Debug-mode) output for the failing boards?

**Comment by shooka90** [comment]
Well something strange happened. I enabled the debug mode feature as you asked, and I can see my packets being received in the Serial monitor and bunch of other stuff that i don't know the meaning of it. Here i post it for you.(DATA from Serial Monitor)

[debugmode_promini5v16mhz.txt](https://github.com/jgromes/RadioLib/files/5964116/debugmode_promini5v16mhz.txt)


The verbos is also here:

[verbos_debugmode_promini5v16mhz.txt](https://github.com/jgromes/RadioLib/files/5964086/verbos_debugmode_promini5v16mhz.txt)

I have to say that my configuration is currently like this: I have an arduino promini 5v 16mhz which I am powering it with 3.3v. I have a Lora module connected to the SPI pins.

One more thing, I disbaled the debugmode mode and then the same thing. I got 
12:52:27.365 -> [SX1278] Initializing ... failed, code -16
in the serial monitor.

Also I have to say, I have an NRF24l01 module. The module doesnt run into any problem with my arduino promini. and my other code for the nrf module works. I guess something is not write in the library.

**Reply by shooka90** [reply]
*typos: write --> right

**Reply by shooka90** [reply]
Also I have checked the board manger, I have the latest version installed. (1.8.3). I have tried to get the board info. Its unknown. So I guess I am using a clone. But its definitely an ATMEGA 328P.

**Reply by jgromes** [reply]
I tried running the code from your previous post on a couple Arduino Pro Mini and NANO boards, but I can't seem to be able to replicate the issue - everything is working fine on my end. I tried both 8 MHz and 16 MHz Pro Mini and a Nano (with a level shifter), same result each time. By accident I also uploaded a code compiled for Nano into the 16 MHz Pro Mini and that worked too.

I used the same setup you have, with Arduino IDE 1.8.13 and AVR core 1.8.3, and the compiled binary I have is the same exact size as yours, so I think the program is identical. So at this point, the only explanation I can think of is an issue with the hardware.

**Comment by shooka90** [comment]
Well, thank you for your time, but some questions remained unanswered:

1- Why when enable debug feature of your library, the module starts to work and receive data and it wont fail, but when I disable the debug feature the module fails? (Error code:-16)

2- I have no problem using my arduino promini with other libraries like RF24. They recognize the SPI correctly and no problem will occure. So why I ran into problems with radiolib library?

I will try to change my arduino promini, and post the results.

**Reply by jgromes** [reply]
To (somewhat) answer those questions, you should understand that RadioLib was explicitly written to be as careful as possible when accessing the hardware. For example, each time there's a write to an SPI register, the newly written value is verified to check the register did actually update. Most other Arduino libraries just perform the write and assume it will always succeed. So if an SPI write fails in RadioLib, you get an error report, while in other libraries you get nothing, making you (the user) think everything is fine.

As to why the issue disappears when using debug mode - that likely has something to do with the fact that using debug mode will slow down the rate of SPI communication. Not its clock frequency, just the rate at which transfers are made. But I really don't know why should that matter. At this point, I would probably grab an oscilloscope and check the SPI communication directly, not sure if that's an option for you.

**Comment by shooka90** [comment]
My new shipment of new arduinos has arrived. I  have tested my case, and everything was fine and no problem occurred. Maybe the faulty arduino spi was the problem. You can officially terminate this discussion.
Thanks for your help.
